tweak UtBS difficulty

cooljeanius opened the pull request:
Like with my previous campaign-related PRs (#5268, #5496, #5497, #5524, #5382, #5603, and #6057), this basically just attempts to apply the changes that I made to the campaign in question (UtBS in this case) during my most recent playthrough of it.
Closes cooljeanius/wesnoth_mods#9
Originally taken from https://github.com/cooljeanius/wesnoth_mods/tree/master/campaigns/Under_the_Burning_Suns/diffs
Note: contains string changes, so it'll have to wait until 1.19 or later.

cooljeanius committed:
initial attempt at applying my patches

see cooljeanius/wesnoth_mods#9
see https://github.com/cooljeanius/wesnoth_mods/tree/master/campaigns/Under_the_Burning_Suns/diffs

cooljeanius committed:
try wiggling in the remaining rejects files

see cooljeanius/wesnoth_mods#9
see https://github.com/cooljeanius/wesnoth_mods/tree/master/campaigns/Under_the_Burning_Suns/diffs

cooljeanius committed:
Recreate 11_Battle_for_Zocthanol_Isle.cfg

(oops, previous commit contained an unintended deletion)

cooljeanius committed:
Update 08_Out_of_the_Frying_Pan.cfg

remove portion causing a validator error

stevecotton commented on data/campaigns/Under_the_Burning_Suns/scenarios/11_Battle_for_Zocthanol_Isle.cfg:
```diff
@@ -180,8 +180,8 @@
-            aggression=0.8
-            caution=0.1
+            aggression={ON_DIFFICULTY 0.6 0.7 0.8}
+            caution={ON_DIFFICULTY 0.3 0.2 0.1}
```
Many places would benefit from comments about what the intended effect is, what play styles is it meant to be more or less lenient for.

cooljeanius replied:
basically here when that side was more aggressive, it was killing me more; the idea is to have them back off a bit on easier difficulties to give the player a bit more breathing room

cooljeanius replied:
comment added in [dd0c083](https://github.com/wesnoth/wesnoth/pull/8421/commits/dd0c08340433d9d125b45e46e014426d46b0fdc3).

stevecotton commented on data/campaigns/Under_the_Burning_Suns/scenarios/06a_In_the_Tunnels_of_Trolls.cfg:
```diff
@@ -929,7 +941,7 @@
         [message]
             speaker=narrator
-            message= _ "Any unit that ends its turn on a lava hex, except those who can fly over the lava, will take 25 damage at the beginning of the next turn. This lava damage can kill units. Flying units will just take $heat_damage damage per turn when flying over lava, though they too can die if they spend too much time over it. Also because of the heat in the cavern, all units on cave floor ...
+            message= _ "Any unit that ends its turn on a lava hex, except those who can fly over the lava, will take $lava_damage damage at the beginning of the next turn. This lava damage can kill units. Flying units will just take $heat_damage damage per turn when flying over lava, though they too can die if they spend too much time over it. Also because of the heat in the cavern, all units on  ...
```
Is the "can't kill" right? It contradicts the comment on line 801.

cooljeanius replied:
both the "can kill" in the comment and the "can't kill" in this string are unchanged from the current revision; I forget which is correct...

cooljeanius replied:
oh wait a moment, I see what's going on here... there are 2 different sources of damage here: the heat damage, which is the one that can't kill units, and the lava damage, which is the one that *can* kill units... maybe this string should be broken up so it's not talking about both in a single string?

cooljeanius replied:
I adjusted the string in [dd0c083](https://github.com/wesnoth/wesnoth/pull/8421/commits/dd0c08340433d9d125b45e46e014426d46b0fdc3), but it might still be worth considering splitting it...

stevecotton replied:
I think splitting is necessary, otherwise it's a wall of text that says "lava damage" at the start and "can't kill" at the end.

cooljeanius replied:
Split in [c63001d](https://github.com/wesnoth/wesnoth/pull/8421/commits/c63001d8dcdf3e4022beeef08f98c644d15db1cb).

stevecotton commented on data/campaigns/Under_the_Burning_Suns/scenarios/06a_In_the_Tunnels_of_Trolls.cfg:
```diff
@@ -689,14 +696,18 @@
             y=4
             side=2
-            hitpoints=20
+            hitpoints={ON_DIFFICULTY 40 30 20}
```
What are these meant to mean for the player?

It's a very tightly scripted sequence anyway, so difficulty shouldn't make much different; however it will make it much more likely for a bug to go undetected during playtesting if something about the sequence changes in a way that makes HARD mode overly dependent on RNG.

cooljeanius replied:
When I played through this scenario, the trolls were sometimes killing Rogrimir before I could save him; giving him more HP is supposed to give the player more time to save him. The HARD mode value is preserved because I assumed that HARD mode players were ok with being overly dependent on RNG

cooljeanius replied:
comment added in [dd0c083](https://github.com/wesnoth/wesnoth/pull/8421/commits/dd0c08340433d9d125b45e46e014426d46b0fdc3).

cooljeanius replied:
Open to further negotiation about the exact HP values, as long as it stays a linear progression with more HP on easier difficulties.

stevecotton commented on data/campaigns/Under_the_Burning_Suns/scenarios/02_Across_the_Harsh_Sands.cfg:
```diff
@@ -1044,7 +1046,7 @@
             profile=portraits/elyssa.webp
             gender=female
-            experience=15
+            experience={ON_DIFFICULTY 17 16 15}
```
Why?

cooljeanius replied:
to make it easier to level her up earlier; I think I had some cases where she died on me just a few XP away from advancing...

stevecotton replied:
For a unit that needs 125 XP, a difference of just 1XP sounds like it would be specific to the RNG that you got.

cooljeanius replied:
ok so maybe `{ON_DIFFICULTY 35 25 15}` then?

Hejnewar replied:
Reminder that XP now scales with difficulty. I get her almost leveled on nightmare with 200% xp mod. Is there some cool interaction if she levels? If not there should be I think.

cooljeanius replied:
> Reminder that XP now scales with difficulty.

It does? When did that happen?

Hejnewar replied:
Well I already said that a lot happened. Perhaps it was not clear enough. You should play it or check commit history. Or both.

demario-w replied:
Why should everyone that needs to know what changed have to browse through the commit history?

Hejnewar replied:
I would assume that knowing the up to date version of content that you want to change is the bare minimum when it comes to making changes. You also don't have to do that, checking scenarios and the game / content i usually s enough as I also stated before.

demario-w replied:
And how exactly are we supposed to know and be sure we know them all, as you are not sharing it in a descriptive way?

It is obvious that the author of this PR has played it as he did actually implement changes to make it workable to him.

Hejnewar replied:
I'm not saying this for whatever your comment is about I'm saying that because author admitted that these were changes for 1.14 which is pretty much 2 major versions in the past bruh. So what you are saying is obvious... Is not true since author stated that himself earlier. XD

Hejnewar replied:
Also if even easier version is needed then perhaps add 4th difficulty.

cooljeanius replied:
> I'm not saying this for whatever your comment is about I'm saying that because author admitted that these were changes for 1.14 which is pretty much 2 major versions in the past bruh.

Just 1; 1.18 still isn't out yet. Also the fact that I was able to get these diffs to apply at all shows that things haven't changed *THAT* much since 1.14...

Hejnewar replied:
Because the scenarios werent reworked, but balance and units were.

cooljeanius replied:
@Hejnewar is this the part where you meant "XP maybe 25 on all difficulties"? I prefer more variation between difficulties rather than less, so how about one of these:
- `{25 20 15}`
- `{35 25 15}`
- `{30 25 20}`
- `{27 26 25}`
- `{26 25 24}`
- `{25 24 23}`
- `{25 23 21}`
- `{25 22 19}`
- `{25 21 17}`

cooljeanius commented:
N.B. I'm keeping this in Draft Mode until I've had a chance to re-test; when I originally made these changes, I was still on 1.14, so I'll have to see if any of the other changes to mainline since then have affected anything...

cooljeanius commented:
> I mean this campaign is already not even normal in my opinion on normal (its easier) but I might be biased.

And meanwhile I thought it was really hard even on easy, so...

demario-w commented:
> I mean this campaign is already not even normal in my opinion on normal (its easier) but I might be biased.
> 
Yep. You're actually so good at this game... that your opinion on difficulty doesn't really matter anymore.

Let's cherish [this artifact](https://replays.wesnoth.org/1.14/2021/01/15/2p__Caves_of_the_Basilisk_Turn_12_(82077).bz2) of a rare public display of your playing skill. rofl

Hejnewar commented:
Ah yes that game I remember it. I had fun there, playing differently like this was actually nice for a change. It wasnt even that far from being in winning possition. It was certainly worth your time to remind me of this nice memory.

But yes skill shaming people you disagree with is beautiful tactic to try and shut them up. So nice of you to expose yourself as such a person right before everyones eyes. But that doesnt change my opinion. On normal (lowest difficulty) I can beat it without thinking.

stevecotton commented:
Gentlemen, could I redirect your energies from arguing into critiquing whether #8417 is too little or too much detail for PO hints?

I asked and got feedback from tamanegi-jp (Japanese) and hermestrismi (Arabic) for the first of the EI hints series, would be good to see if I've got it right or wrong on the later ones.

cooljeanius committed:
address some things noticed in review

ePaul commented on data/campaigns/Under_the_Burning_Suns/scenarios/02_Across_the_Harsh_Sands.cfg:
```diff
@@ -1430,17 +1432,42 @@
-
+#endif
         {CHECK_EXPLORER}
+#ifdef EASY
+        [message]
+            speaker=$explorer.id
+            message= _ "Phew, thank goodness that this oasis is here! After hearing about the bandits in this area, I was beginning to worry that they had created a mirage to trick thirsty travelers such as ourselves!"
+        [/message]
+        [message]
+            speaker=Nym
+            message= _ "I know these bandits are evil, but do you really think they'd stoop so low as to get their victims' hopes up before crushing them?"
+            image=portraits/nym_moody.png
+        [/message]
+        [message]
+            speaker=Zhul
+            message= _ "It has been known to happen before."
+        [/message]
+        [message]
+            speaker=Kaleh
+            message= _ "Truly, there are some depraved and cruel individuals in this world."
+        [/message]
+#else
```
I think this kind of dialog makes more sense when we actually find it's a mirage, instead of when we don't have one.

cooljeanius replied:
also, it's important to put *something* there to explain why the mirage thing isn't happening, when players might expect it to be

ePaul replied:
Okay, that explains why it's there fore EASY, but it might make sense to have a similar dialogue (a bit less hypothetical) also for the other difficulties – i.e. when you find it's a mirage, do some speculation whether the bandits have set this up intentionally.

cooljeanius replied:
Added some additional dialogue in [66a70fa](https://github.com/wesnoth/wesnoth/pull/8421/commits/66a70fad038af1e557e864037c8ca6cd48a7c7e6).

cooljeanius committed:
Merge branch 'wesnoth:master' into egallager_UtBS_difficulty

cooljeanius committed:
Update 02_Across_the_Harsh_Sands.cfg

add some additional dialogue as requested by @ePaul

cooljeanius committed:
Update 04_Descending_into_Darkness.cfg

`wmllint` incorrectly wanted to remove a translation mark, so prevent it from doing so

cooljeanius committed:
Update 06a_In_the_Tunnels_of_Trolls.cfg

split damage notice into 2 messages with separate strings, per request from @stevecotton in review

Wedge009 commented on data/campaigns/Under_the_Burning_Suns/scenarios/02_Across_the_Harsh_Sands.cfg:
```diff
@@ -1430,17 +1432,55 @@
+        [/message]
+        [message]
+            speaker=Nym
+            message= _ "I know these bandits are evil, but do you really think they'd stoop so low as to get their victims' hopes up before crushing them?"
```
Typographic apostrophes needed.

cooljeanius replied:
...right, fixed in [1a39c06](https://github.com/wesnoth/wesnoth/pull/8421/commits/1a39c060ce6fb1d7ec7d61d79e63d7c106a45a60).

cooljeanius committed:
Update 02_Across_the_Harsh_Sands.cfg

use typographic apostrophes in dialogue

Hejnewar commented:
Is this still relevant in any way right now?

cooljeanius commented:
> Is this still relevant in any way right now?

I mean, it still applies, it's just that no one's tested it recently...

Hejnewar commented:
Can you give me like TLDR of the changes?

cooljeanius commented:
> Can you give me like TLDR of the changes?

See: https://github.com/cooljeanius/wesnoth_mods/blob/master/campaigns/Under_the_Burning_Suns/README.txt

Hejnewar commented:
S2:
Turns are fine
XP maybe 25 on all difficulties
I do think that dehydration could be better (at the same time I really cheese this scenario in multiple ways)
@Dalas121 for dialogue review

S4: 
I dont think easy level player needs that much gold even, I mean it is as much as all of the AI start with. I would rather see 250/225/200 for this scenario as people say it is a bit too hard on nightmare.
Income is fine tho. 

S6a:
Turns are fine
Income is fine
Ai in my opinion is not needed. 
Lava is fine
Dwarf buff... Hard to say never happened to me and @stevecotton had some worries.
@Dalas121 for dialogue review

S8:
Just delete the unused part, comment can stay for future generations xd
Making it easier on easy is fine. 
But I dont think making it harder on nightmare is needed. 
Nor making it easier so timer of 9/7/5. 

S9: nothing much happens

S11:
I dont think using aggression and caution is a good idea, I would rather use micro ai and gold as a way to change that. 
Other than that it looks fine.

S12 and boss units: 
Is fine. But on nightmare it should stay as 11 damage for the minion. 

At the same time I would like to make this scenario longer / harder / **phase 2** (despite bearly beating it myself on current nightmare cuz of underleveling heh) since while well designed it isnt epic enough.

cooljeanius committed:
Update 08_Out_of_the_Frying_Pan.cfg

Apply suggestions from @Hejnewar

Hejnewar commented:
> You underestimate the difficulty that easy-level players face; we are often worse than you expect.

No, You underestimate Me! More of my losses comes from easy difficulty than from nightmare one! Also this is expert campaign, this is basically easiest easy out of any expert campaign I think and it is easier than it was in the past as well. 

> How about 260/230/200?

Barely doing anything but if it makes you happy. 

> Could you be a bit more specific about what sort of MAI you'd mean?

Im probably had something in mind back then, but tbh this is pretty weak scenario imo. Could be better. 

> 5/8/11 then? 

Tbh it probably would be best to just remove that in the future. I dont really want the fight to be a pushover even on easy. 

> I think @knyghtmare was working on this; see https://github.com/wesnoth/wesnoth/issues/6372 and https://github.com/wesnoth/wesnoth/issues/6142

Also see https://github.com/wesnoth/wesnoth/issues/9459

cooljeanius committed:
Update 04_Descending_into_Darkness.cfg

new negotiated gold amount

cooljeanius commented:
> > You underestimate the difficulty that easy-level players face; we are often worse than you expect.
> 
> No, You underestimate Me! More of my losses comes from easy difficulty than from nightmare one! Also this is expert campaign, this is basically easiest easy out of any expert campaign I think and it is easier than it was in the past as well.
> 
> > How about 260/230/200?
> 
> Barely doing anything but if it makes you happy.

OK, done in c04a34e

> 
> > Could you be a bit more specific about what sort of MAI you'd mean?
> 
> Im probably had something in mind back then, but tbh this is pretty weak scenario imo. Could be better.
>

I think doing too much more of a rework would be out-of-scope for this particular PR...

> > 5/8/11 then?
> 
> Tbh it probably would be best to just remove that in the future. I dont really want the fight to be a pushover even on easy.
>

OK, so can we have a separate "pushover" difficulty then? Some players just want to see the story without being challenged at all.

> > I think @knyghtmare was working on this; see #6372 and #6142
> 
> Also see #9459

Hejnewar commented:
> OK, so can we have a separate "pushover" difficulty then? Some players just want to see the story without being challenged at all.

We already have it, its called debug mode. Why try to play in the first place if you dont want to play, is it a book? I mean it can be help entry I guess.

cooljeanius commented:
> > OK, so can we have a separate "pushover" difficulty then? Some players just want to see the story without being challenged at all.
> 
> We already have it, its called debug mode. Why try to play in the first place if you dont want to play, is it a book?

The thing that makes me prefer video games to books is merely interactivity, not challenge. You can have interactivity without challenge. I like clicking on stuff and seeing it react to me.

Hejnewar commented:
> The thing that makes me prefer video games to books is merely interactivity, not challenge. You can have interactivity without challenge. I like clicking on stuff and seeing it react to me.

Most of wesnoth is unit combat, that takes the most time, in fact it takes longer for me on easier difficulties with higher gold becasue there simply is more units. Why do you (or other person that just wants the stroy) subject yourself to time you dont enjoy (which is majority of the game)? Im pretty sure you can read most of the story on wiki already, dialogues? Thats a bit harder but well someone can do it. Ultimetly IMO this is just simply not a visual novel, it is turn based strategy game.

cooljeanius commented:
> > The thing that makes me prefer video games to books is merely interactivity, not challenge. You can have interactivity without challenge. I like clicking on stuff and seeing it react to me.
> 
> Most of wesnoth is unit combat, that takes the most time, in fact it takes longer for me on easier difficulties with higher gold becasue there simply is more units.

You don't *have* to use all the gold available. You can just recruit fewer units if you feel there are too many.

> Why do you (or other person that just wants the stroy) subject yourself to time you dont enjoy (which is majority of the game)? Im pretty sure you can read most of the story on wiki already, dialogues?

I do enjoy clicking on units and moving them, I don't enjoy losing. Reading the story on the wiki is not an equivalent experience because there is no interactivity.

> Thats a bit harder but well someone can do it. Ultimetly IMO this is just simply not a visual novel, it is turn based strategy game.
> 
> How many people come here to have the visual novel experience if we literally market ourselves as TBS game?
>

There are plenty of players who have heard "Oh, Wesnoth is basically a FOSS version of Fire Emblem", and the Fire Emblem series includes difficulty settings for fans of visual novels.

Hejnewar commented:
> You don't have to use all the gold available. You can just recruit fewer units if you feel there are too many.

Ok, why do we have difficulty levels at all then? Why not just have one with infinite gold? Surely players will just recruit less if they want harder challenege. 

> I don't enjoy losing.

You dont enjoy thinking how to win? Losing is the best, it makes you think, thinking makes you improve, improving makes you overcome the challenge. Not without reason Quenoth after rework have insanely high skill cap and people like it and enjoy it. 

> There are plenty of players who have heard "Oh, Wesnoth is basically a FOSS version of Fire Emblem", and the Fire Emblem series includes difficulty settings for fans of visual novels.

But this campaign doesnt have one and is gated behind scenario 1 where you dont recruit at all. If you cant beat it on easy maybe this campaign isnt for you, just like not being able to beat it on nightmare indicates that that difficulty isnt for that player.

So far the feedback for this campiagn was great and no one except for you (who didnt even play it after changes) considers it to be too hard on easy. Some scenarios might be too hard on nightmare thats true but the target audience for this campaign were people who enjoy the challenege. And it was stunning success for me. Having easy be too easy simply does not align with my vision for this campaign.

ForestDragon-wesnoth commented:
Well, people looking a challenge usually don't click on lowest difficulty, people usually click whatever is in the middle. And considering UTBS is very wildly different from the rest of mainline, for better or worse, not much harm in making the lowest difficulty a bit easier for those interested in trying it

Hejnewar commented:
Nope, I literally agreed to partially agreed to most stuff here without particular need because this was targeted at 1.16 version not 1.18. Im not changing this one. The easy mode is already easier than it was in 1.16. Boss fight is a boss fight its the end of the campaign, if you got this far a bit of a challenge wont kill you and if you didnt but you are still here then you know how to use debug mode.

If you want I will just rename it from easy to normal and make it a bit harder to not confuse players. Ultimately, because it has so many mechanics starting form S2 new player shouldnt really play it firts anyway because most likely they wont know what hit them (and why their units have no attack) and this will just be a bad experience which is something we shouldnt promote. 

Also did any of you actually test 1.18 version on easy difficulty? If yes I would like to get feedback for it.

cooljeanius committed:
Update Alien_Minion.cfg

put Nightmare damage back to 11

cooljeanius commented:
Fine, I put the damage on Nightmare back to 11. I still want to keep my reductions to the Easy and Normal difficulties, though.

Hejnewar commented:
This campaign starts from **challening** difficulty lmao. Now I say its too easy. Because it has formally no easy difficulty. Thats actually a blunder from me now Im literally marketing the difficulty wrong... Looking at the names I really should make all difficulties except for nightmare harder (yes that one too but as nightmare+). 

Right now on easy you need either 3 lvl 3 units or lvl +-5 (thats low I had them on way higher lvl on nightmare with 200% exp mod) heroes (as thats the fastest possible power spike) and some random units. Both heroes and lvl 3 units in most cases are stronger than they were during 1.16. I will ask a question, did you even test these changes on 1.18?  

So my points are:
1. This is not even easy difficulty it is challenging why should it be easy?
2. This is not campaign for new players, it shouldnt invite them, they will lose on S2 (if not S1) even with infinite gold. 
3. Both heroes and units are stronger than in times of 1.16. 
4. The last scenario is on the chopping block already anyway. Honestly I could just say fine here because its going to be changed anyway so it functionally doesnt matter.

cooljeanius committed:
Update Alien_Minion.cfg

put NORMAL value for fangs damage back to 9

cooljeanius commented:
OK, I put the NORMAL value back to its previous value, too, I'm still keeping EASY at 6, though.

cooljeanius commented:
@stevecotton now that Hejnewar has approved, can you take another look at the parts where you had review comments?

Dalas121 commented on data/campaigns/Under_the_Burning_Suns/scenarios/02_Across_the_Harsh_Sands.cfg:
```diff
@@ -1430,17 +1432,55 @@
         [/message]
+        [message]
+            speaker=Kaleh
+            message= _ "Perhaps it was the doing of those bandits..."
```
I think it'd be best to leave out these explanations for the oasis. A big part of UtBS's story (and especially this scenario) is the natural hazards of a hard world. Replacing a natural mirage with a bandit-made trap detracts from that.

If the oasis vanishes, $explorer.id can say it must've been a mirage. If the oasis is there, $explorer.id can celebrate its presence.

cooljeanius replied:
I added that part at the request of @ePaul's review comment above...

Dalas121 replied:
I'm not seeing ePaul's comment, but in my opinion the event would be improved by removing all dialogue except _"Dang. I was sure I saw an oasis here. Must have been a mirage. I’ve been out in the sand for too long."_

ePaul replied:
The [original discussion](https://github.com/wesnoth/wesnoth/pull/8421#discussion_r1588520489).

Before my comment, there was some dialog in the EASY mode for the purpose of lampshading "in the higher difficulty modes, the oasis might be a mirage". I said there should rather be something when there is actually a mirage, but accepted the argument that there should still be something when the oasis is real.

I agree with your argument that the whole "bandits might create mirages to lead travelers into traps" idea might be a bit strange, so maybe skipping this and just saying "I'm happy this is not a mirage" (indicating there actually might be (natural) mirages here) sounds better.

Dalas121 replied:
Got it, thanks. I'm fine having easy-difficulty dialogue to lampshade the higher-difficulty mirages, like what you've suggested.

cooljeanius replied:
ok, does [19e7a88](https://github.com/wesnoth/wesnoth/pull/8421/commits/19e7a88db350e45eb338cb178829ff5e3b782ea3) make it better?

cooljeanius committed:
Update 02_Across_the_Harsh_Sands.cfg

adjust dialogue according to review comments

Hejnewar commented:
Ig, this is good for merge.

stevecotton commented:
I'm taking a break from Wesnoth - I will return, but other stuff is taking my time, so please don't wait for me.

I still feel that some of the `ON_DIFFICULTY` additions shouldn't be made for the reasons already commented, but I'm not participating enough to have a voice at the moment.

Hejnewar commented:
Closing this PR as most of the things from it were implemented in one way or another or are no longer relevant in the face of changes.
